### ISSUE TITLE
Optional device wide synchronization to allow parallel persistent or long running kernels

### ISSUE THREAD

**Issue by Evotica** [ultralytics/ultralytics#23678]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

As discussed in https://github.com/ultralytics/ultralytics/issues/21411#issuecomment-3640723278,
I would like see a feature that disables any device wide synchronization calls in Ultralytics code, such as in ultralytics.utils.ops.Profile class.

If some other code used in the same process where Ultralytics code is also used starts a persistent parallel GPU kernel, then any device wide syncronization calls will result in a deadlock of the caller.

### Use case

1. You are using a library, that passes realtime data coming from a NIC directly into GPU memory, for example by using DOCA GPUNetIO or DPDK Gpudev.
2. You have a persistent kernel processing that data into image chunks.
3. You want to run inference with these image chunks, in parallel to that network data handling.

If all this is in the same process, the inference part breaks, if it calls for a device wide synchronization, as any persistent kernels will prevent the synchronization from happening.

### Additional

To bypass the device wide synchronization in ultralytics.utils.ops.Profile class, I have been using the following code to overwrite the ops.Profile.time() method after importing.

```python
# Do this early, before creating the model
# Monkey‑patch the Profile class.
# Original calls for device-wide synchronization that halts the whole system when there are tail lauching kernels running
from ultralytics.utils import ops
class NoSyncProfile(ops.Profile):
    def time(self):                       # override to avoid device sync
        return time.perf_counter()

ops.Profile = NoSyncProfile
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Evotica, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This is an automated response 🤖—an Ultralytics engineer will also assist soon. Thanks for the detailed feature request and the monkey-patch workaround you shared 🙏

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Even though this is a feature request, an MRE that demonstrates the deadlock behavior (plus the exact call stack leading into `ultralytics.utils.ops.Profile`) would be super helpful 🧩

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the detailed write-up—this is a valid concern: `ultralytics.utils.ops.Profile` currently does a device-wide `torch.cuda.synchronize()` whenever it’s constructed with a CUDA device, which can indeed stall/deadlock when persistent kernels are running on the same GPU (see the current implementation in the [Profile docs](https://docs.ultralytics.com/reference/utils/ops/)). We’ll look at adding an official opt-out (e.g. a `sync=False`/env-flag) and potentially switching timing to CUDA events to avoid device-wide sync; for now, your monkey-patch is the right workaround since `BasePredictor.stream_inference()` instantiates `Profile(device=self.device)` internally (reference: [stream_inference() docs](https://docs.ultralytics.com/reference/engine/predictor/)). If you can share which Ultralytics entrypoint triggers the deadlock in your setup (`predict()`, `track()`, etc.) and whether the persistent kernel is on the default stream or a separate stream, that will help us design the safest switch.
